### PR TITLE
[Cocoa] Small caps should not be synthesized if _any_ character in the font supports small caps

### PR DESCRIPTION
**Pull request by litherum**
#### 1946d6148275697aeef24a8049e85848e7ca5136
<pre>
[Cocoa] Small caps should not be synthesized if _any_ character in the font supports small caps
<a href="https://bugs.webkit.org/show_bug.cgi?id=244236">https://bugs.webkit.org/show_bug.cgi?id=244236</a>
&lt;rdar://problem/99004636&gt;

Reviewed by Tim Nguyen.

Previously, we were synthesizing small caps if any character in the run didn&apos;t support small caps.
However, the [spec](<a href="https://drafts.csswg.org/css-fonts-4/#font-variant-caps-prop)">https://drafts.csswg.org/css-fonts-4/#font-variant-caps-prop)</a> says:

&gt; The availability of these glyphs is based on whether a given feature is defined or not in the feature
&gt; list of the font. User agents ... should explicitly not decide this on a per-character basis.

This patch implements that. The new policy is that if a font supports _any_ small caps characters, then
we don&apos;t synthesize any runs with that font.

Tests: imported/w3c/web-platform-tests/css/css-fonts/font-variant-01.html
       imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps-03.html
       imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps-05.html
       imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps.html

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::shouldSynthesize):
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::variantCapsSupportedForSynthesis const):
(WebCore::Font::variantCapsSupportsCharacterForSynthesis const): Deleted.
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::variantCapsSupportedForSynthesis const):
(WebCore::Font::supportsSmallCaps const):
(WebCore::Font::supportsAllSmallCaps const):
(WebCore::Font::supportsPetiteCaps const):
(WebCore::Font::supportsAllPetiteCaps const):
(WebCore::Font::variantCapsSupportsCharacterForSynthesis const): Deleted.
(WebCore::Font::glyphsSupportedBySmallCaps const): Deleted.
(WebCore::Font::glyphsSupportedByAllSmallCaps const): Deleted.
(WebCore::Font::glyphsSupportedByPetiteCaps const): Deleted.
(WebCore::Font::glyphsSupportedByAllPetiteCaps const): Deleted.
* Source/WebCore/platform/graphics/freetype/SimpleFontDataFreeType.cpp:
* LayoutTests/TestExpectations:
* LayoutTests/css3/font-variant-synthesis-jdaggett.html: Replaced by imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps.html
* LayoutTests/css3/font-variant-synthesis-jdaggett-expected.html:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/platform/gtk/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/253819@main">https://commits.webkit.org/253819@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/9cf339f1e77b3630fb7bafdf6a5c84d86b88d13c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87179 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18009 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96197 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149752 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91155 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29630 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91191 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23933 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73986 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/79300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78912 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79159 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27355 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12995 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27302 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14011 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2695 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28987 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/36867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28926 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33288 "Passed tests") | |

### CODE REVIEW

**Comment by nt1m** [comment]
The fix makes sense to me, thanks!

* GTK seems to have trouble building

* css3/font-variant-synthesis-jdaggett.html is very similar to imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps.html. I would delete the former. (They're both made by the same author)

* Also I think you should be able to get rid of these 3:
https://searchfox.org/wubkat/rev/340573c924b370891b7a0794ef39d4995d3670ee/LayoutTests/TestExpectations#3718-3720 (and potentially more, can you re-run `imported/w3c/web-platform-tests/css/css-fonts/` locally to check? EWS doesn't catch unexpected passes).

**Comment by litherum** [comment]
> imported/w3c/web-platform-tests/css/css-fonts/font-variant-01.html (though this isn't marked as passing yet because another fix is necessary on top of this patch)

I think this is no longer true, and I can mark it as passing. See https://bugs.webkit.org/show_bug.cgi?id=244219

File: LayoutTests/TestExpectations [hunk 1]
```diff
@@ -3696,7 +3696,6 @@ webkit.org/b/206881 imported/w3c/web-platform-tests/css/css-fonts/font-kerning-0
 webkit.org/b/206881 imported/w3c/web-platform-tests/css/css-fonts/font-kerning-05.html [ ImageOnlyFailure ]
 webkit.org/b/206881 imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-001.html [ ImageOnlyFailure ]
 webkit.org/b/206881 imported/w3c/web-platform-tests/css/css-fonts/font-size-adjust-002.html [ ImageOnlyFailure ]
-webkit.org/b/206881 imported/w3c/web-platform-tests/css/css-fonts/font-variant-01.html [ ImageOnlyFailure ]
 webkit.org/b/206881 imported/w3c/web-platform-tests/css/css-fonts/font-variant-02.html [ ImageOnlyFailure ]
```
Comment (nt1m) [inline LayoutTests/TestExpectations]: I wonder what the remaining failure is about for this test.

**Review by nt1m — APPROVED** [review]
Thanks!

**Comment by nt1m** [comment]
GTK needs a bit of gardening.

**Pull request merged**